Привет! меня зовут Люман Аблаев. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на красные комментарии ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:


<div class="alert alert-block alert-success">
<b>Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>

-------------------

Будет очень хорошо, если ты будешь помечать свои действия следующим образом:
<div class="alert alert-block alert-warning">
<b>Комментарий студента:</b> ..
</div>
<div class="alert alert-block alert-warning">
<b>Изменения:</b> Были внесены следующие изменения ...
</div>

<font color='orange' style='font-size:24px; font-weight:bold'>Общее впечатление</font>
* Спасибо за  качественную работу, было интересно ее проверять.
- Я оставил некоторые советы надеюсь они будут полезными или интересными.

- Есть небольшие недочеты, но думаю у тебя не составить труда их исправить
- Жду обновленную работу!

# Выбор локации для скважины

Допустим, вы работаете в добывающей компании «ГлавРосГосНефть». Нужно решить, где бурить новую скважину.

Вам предоставлены пробы нефти в трёх регионах: в каждом 10 000 месторождений, где измерили качество нефти и объём её запасов. Постройте модель машинного обучения, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Проанализируйте возможную прибыль и риски техникой *Bootstrap.*

Шаги для выбора локации:

- В избранном регионе ищут месторождения, для каждого определяют значения признаков;
- Строят модель и оценивают объём запасов;
- Выбирают месторождения с самым высокими оценками значений. Количество месторождений зависит от бюджета компании и стоимости разработки одной скважины;
- Прибыль равна суммарной прибыли отобранных месторождений.

## Загрузка и подготовка данных

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split 
from sklearn.metrics import accuracy_score
from sklearn.metrics import mean_squared_error 

<div class="alert alert-block alert-success">
<b>Успех:</b> Импорты  на месте
</div>


In [2]:
try:
    df1 = pd.read_csv('/datasets/geo_data_0.csv', sep = ',')
    df2 = pd.read_csv('/datasets/geo_data_1.csv', sep = ',')
    df3 = pd.read_csv('/datasets/geo_data_2.csv', sep = ',')
except:
    df1 = pd.read_csv('https://code.s3.yandex.net/datasets/geo_data_0.csv', sep = ',')
    df2 = pd.read_csv('https://code.s3.yandex.net/datasets/geo_data_1.csv', sep = ',')
    df3 = pd.read_csv('https://code.s3.yandex.net/datasets/geo_data_2.csv', sep = ',')

df1.info()
df1.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

In [3]:
df2.info()
df2.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

In [4]:
df3.info()
df3.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

В приведённых таблицах нет пропусков и названия столбцов соотвествуют правилам, поэтому данные в предобработке не нуждаются. 

<div class="alert alert-block alert-success">
<b>Успех:</b> Данные осмотрены бегло и можно было бы еще как минимум изучить корреляции, так как мы планируем использовать линейные модели, а для них это важно. 
</div>


<div class="alert alert-block alert-info">
<b>Совет:</b>  Советую обратить внимание на библиотеки sweetviz и pandas_profiling помогут в проведении более тщательного EDA анализа. Исследовательский анализ можно делать и с помощью ручного вызова функций дефолтных библиотек. Данные библиотеки хороши для максимизации комфорта презентации результатов анализа бизнес-пользователям.  

Очень просты в использоовании, на вход кладется датафрейм: pandas_profiling.ProfileReport(df) 
</div>


## Обучение и проверка модели

Разделим все 3 датафрейма на обучающую и валидационную выборки в соотношении 75:25:

In [5]:
features1 = df1.drop(['product', 'id'], axis = 1)
target1 = df1['product'] 

features_train1, features_valid1, target_train1, target_valid1 = train_test_split(
    features1, target1, test_size=0.25, random_state=12345)

model = LinearRegression()
model.fit(features_train1, target_train1) 
predictions_valid1 = model.predict(features_valid1) # предсказания модели на валидационной выборке

mid1 = predictions_valid1.mean() # средний запас предсказанный на валидационной выборке
mid1 = float('{:.3f}'.format(mid1)) # округлим результат до 3 цифр после запятой

result1 = mean_squared_error(target_valid1, predictions_valid1) ** 0.5 # RMSE
result1 = float('{:.3f}'.format(result1)) # также округлим
print("Средний запас предсказанного сырья и RMSE модели линейной регрессии на валидационной выборке:", mid1, '+-', result1)

Средний запас предсказанного сырья и RMSE модели линейной регрессии на валидационной выборке: 92.593 +- 37.579


Для первого региона RMSE довольно большой относительно среднего, что не очень хорошо, теперь проверим эти значения для второго региона:

In [6]:
features2 = df2.drop(['product', 'id'], axis = 1)
target2 = df2['product'] 

features_train2, features_valid2, target_train2, target_valid2 = train_test_split(
    features2, target2, test_size=0.25, random_state=12345)

model = LinearRegression()
model.fit(features_train2, target_train2) 
predictions_valid2 = model.predict(features_valid2) # предсказания модели на валидационной выборке

mid2 = predictions_valid2.mean() # средний запас предсказанный на валидационной выборке
mid2 = float('{:.3f}'.format(mid2)) # округлим результат до 3 цифр после запятой

result2 = mean_squared_error(target_valid2, predictions_valid2) ** 0.5 # RMSE
result2 = float('{:.3f}'.format(result2)) # также округлим
print("Средний запас предсказанного сырья и RMSE модели линейной регрессии на валидационной выборке:", mid2, '+-', result2)

Средний запас предсказанного сырья и RMSE модели линейной регрессии на валидационной выборке: 68.729 +- 0.893


Для второго региона средний запас сырья меньше, но RMSE гораздо меньше, что говорит о малой среднеквадратичной ошибке.

In [7]:
features3 = df3.drop(['product', 'id'], axis = 1)
target3 = df3['product'] 

features_train3, features_valid3, target_train3, target_valid3 = train_test_split(
    features3, target3, test_size=0.25, random_state=12345)

model = LinearRegression()
model.fit(features_train3, target_train3) 
predictions_valid3 = model.predict(features_valid3) # предсказания модели на валидационной выборке

mid3 = predictions_valid3.mean() # средний запас предсказанный на валидационной выборке
mid3 = float('{:.3f}'.format(mid3)) # округлим результат до 3 цифр после запятой

result3 = mean_squared_error(target_valid3, predictions_valid3) ** 0.5 # RMSE
result3 = float('{:.3f}'.format(result3)) # также округлим
print("Средний запас предсказанного сырья и RMSE модели линейной регрессии на валидационной выборке:", mid3, '+-', result3)

Средний запас предсказанного сырья и RMSE модели линейной регрессии на валидационной выборке: 94.965 +- 40.03


У последнего региона самое большое среднее, но также и самый большой корень из среднеквадратичной ошибки.



<div class="alert alert-block alert-success">
<b>Успех:</b> Шаг проделан хорошо, все метрики посчитаны корректно
</div> 


<div class="alert alert-block alert-info">
<b>Совет: </b>  Можно было бы использовать циклы/функции и избежать дублирования кода. Дублирование кода, часто приводит к глупым ошибкам, которые тяжело отследить.
</div>


## Подготовка к расчёту прибыли

Для начала предлагаю все считать в млрд рублей просто из соображений удобства, тогда основные величины будут следующими:

In [8]:
TOTAL_COSTS = 10 # расходы на разработку скважин
PRICE_BAR = 450000/(10**9) # цена единицы продукта
NUM = 200 # кол-во скважин для разработки


<div class="alert alert-block alert-info">
    
<b>Совет:</b> Константы принято записывать в ВЕРХНЕМ РЕГИСТРЕ =) Читем здесь https://pythonworld.ru/osnovy/pep-8-rukovodstvo-po-napisaniyu-koda-na-python.html#section-29
    
А большие значение можно записывать так: `450_000`.
</div> 


<div class="alert alert-block alert-warning">
<b>Изменения:</b> Переделал названия констант.
</div>

<div class="alert alert-block alert-success">
    
<b>Успех[2]:</b> 👍
</div>


Теперь рассчитаем сколько единиц продукта в среднем должна добывать скважина для безубыточного производства:

In [9]:
MIN_MEAN_PRODUCT = TOTAL_COSTS / (PRICE_BAR * NUM)
MIN_PRODUCT = TOTAL_COSTS / PRICE_BAR
print('Минимальное сырьё со всех скважин', MIN_PRODUCT)
print('Минимальное сырьё в среднем с одной скважины', MIN_MEAN_PRODUCT)

Минимальное сырьё со всех скважин 22222.222222222223
Минимальное сырьё в среднем с одной скважины 111.11111111111111


Чуть больше 111 единиц продукта должна добывать в среднем скважина, поэтому сразу можно сделать вывод, что 2 регион не подходит, 1 и 3 регионы входят в этот промежуток с погрешностями RMSE.



<div class="alert alert-block alert-success">
<b>Успех:</b> Значение для безубыточной разработки посчитано верно, с выводом согласен  
</div>

Напишем функцию расчёта прибыли, учитывая предыдущие показатели:

In [10]:
def profit(target, predictions):
    target.reset_index(drop= True , inplace= True)
    predictions.reset_index(drop= True , inplace= True)
    predictions_sorted = predictions.sort_values(ascending=False)[:NUM] # оставляем 200 лучших скважин по добыче сырья
    selected = target[predictions_sorted.index]
    return (selected.sum() - MIN_PRODUCT)  * PRICE_BAR # складываем всё сырьё и вычитаем из минимального сырья 
# и умножаем эту разницу на стоимость единицы сырья

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Для выбора 200 лучших скважин нужно использовать предсказанные значения. А вот для подсчета прибыли должны использоваться настоящие значения соответствующих ячеек.

</div>

<div class="alert alert-block alert-warning">
<b>Изменения:</b> Теперь суммируются настоящие значения, а не предсказания. Цикл во 2 и 4 пунктах действительно недодумался сделать, но переделывать это довольно долго, хотелось бы приступить к сборному проекту, так как немного отстаю) На будущее учту)
</div>

<div class="alert alert-block alert-danger">
    
<b>Ошибка[2]:</b> Да, теперь ф-ция определна логически верно, и я бы не хотел тебя задерживать, но увы  есть 1 неявный  и очень интересный недочет, на котороый я хотел бы обратить твое внимание: из-за того, что у нас техника бустрап выполняется с возвращением (replace=True), то могут отобраться скважины с одинаковым индексом, и тогда при подстановки предсказаний в таргеты мы получим не 200 скважин, а 200+ и прибыль получится завышенной (можешь вывести длину таргетов после подстановки - и убедишься в этом) 

    
Тут нужно сбросить индексы для входных данных в начале функции прибыли. Или собрать таргеты и предсказания в датафрейм - тогда они и сэмплироваться и сортироватсья будут как единное целое.
</div>


## Расчёт прибыли и рисков 

С помощью Bootstrap оценим доверительный интервал в 0.95% для 1 и 3 региона:

In [11]:
state = np.random.RandomState(12345)

target_valid1.reset_index(drop= True , inplace= True)
predictions_valid1 = pd.Series(predictions_valid1)
   
values = []
k = 0
for i in range(1000):
    # по условию из 500 скважин выбираются наилучшие по сырью, поэтому n = 500
    target_valid_subsample1 = target_valid1.sample(n=500, replace=True, random_state=state)
    predictions_valid_subsample1 = predictions_valid1[target_valid_subsample1.index]
    
    
    values.append(profit(target_valid_subsample1, predictions_valid_subsample1))
    if profit(target_valid_subsample1, predictions_valid_subsample1) < 0: # если прибыль отрицательная увеличиваем счётчик
        k += 1

bad_probability = (k / 1000) * 100 # вероятность убытка
values = pd.Series(values)
upper = values.quantile(0.975)
lower = values.quantile(0.025)

mean = values.mean()
print("Средняя выручка:", mean)
print("97.5%-квантиль:", upper)
print("2.5%-квантиль:", lower)
print('Вероятность убытка:', bad_probability, '%')

Средняя выручка: 0.396164984802371
97.5%-квантиль: 0.9097669415534224
2.5%-квантиль: -0.11121554589049622
Вероятность убытка: 6.9 %


<div class="alert alert-block alert-success">
<b>Успех:</b> С бутстрапом  все в порядке, метрики считаются верным образом!  
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b>
    
Риски можно лаконично считать в 1 строку: `(values < 0).mean() * 100`
</div>          

Для 1 региона вероятность убытка  - 6.9 %, что больше 2.5%, а средняя выручка в районе 0.4 млрд. Посмотрим на прибыль для 3 региона:

In [12]:
state = np.random.RandomState(12345)

target_valid3.reset_index(drop= True , inplace= True)
predictions_valid3 = pd.Series(predictions_valid3)
    
values = []
k = 0
for i in range(1000):
    # по условию из 500 скважин выбираются наилучшие по сырью, поэтому n = 500
    target_valid_subsample3 = target_valid3.sample(n=500, replace=True, random_state=state)
    predictions_valid_subsample3 = predictions_valid3[target_valid_subsample3.index]
    
    values.append(profit(target_valid_subsample3, predictions_valid_subsample3))
    if profit(target_valid_subsample3, predictions_valid_subsample3) < 0: # если прибыль отрицательная увеличиваем счётчик
        k += 1

bad_probability = (k / 1000) * 100 # вероятность убытка
values = pd.Series(values)
upper = values.quantile(0.975)
lower = values.quantile(0.025)

mean = values.mean()
print("Средняя выручка:", mean)
print("97.5%-квантиль:", upper)
print("2.5%-квантиль:", lower)
print('Вероятность убытка:', bad_probability, '%')

Средняя выручка: 0.40440386656835664
97.5%-квантиль: 0.9503595749238002
2.5%-квантиль: -0.16335041339560044
Вероятность убытка: 7.6 %



<div class="alert alert-block alert-info">
<b>Совет: </b>
   
Здесь циклы/функции тоже были бы весьма кстати
</div>

Тут вероятность убытка выше, чем в первом регионе, а средняя прибыль чуть больше.

На всякий случай проверим и 2 регион:

In [13]:
state = np.random.RandomState(12345)

target_valid2.reset_index(drop= True , inplace= True)
predictions_valid2 = pd.Series(predictions_valid2)
    
values = []
k = 0
for i in range(1000):
    # по условию из 500 скважин выбираются наилучшие по сырью, поэтому n = 500
    target_valid_subsample2 = target_valid2.sample(n=500, replace=True, random_state=state)
    predictions_valid_subsample2 = predictions_valid2[target_valid_subsample2.index] 
    
    values.append(profit(predictions_valid2, predictions_valid_subsample2))
    if profit(target_valid2, predictions_valid_subsample2) < 0: # если прибыль отрицательная увеличиваем счётчик
        k += 1

bad_probability = (k / 1000) * 100 # вероятность убытка
values = pd.Series(values)
upper = values.quantile(0.975)
lower = values.quantile(0.025)

mean = values.mean()
print("Средняя выручка:", mean)
print("97.5%-квантиль:", upper)
print("2.5%-квантиль:", lower)
print('Вероятность убытка:', bad_probability, '%')

Средняя выручка: -3.9749035884105286
97.5%-квантиль: -3.536493095438133
2.5%-квантиль: -4.435856538392783
Вероятность убытка: 100.0 %


У второго региона вероятность убытка 100%, а значит предыдущий вывод, что он не подходит по средней добыче - верен. По всем результатам выберем 1 регион, так как там самая малая вероятность убытка (но все равно выше 2.5%), а среднее примерно такое же как в 3 регионе.



<div class="alert alert-block alert-info">
<b>Совет: </b>
    
Не забудь обновить выводы, если потребуется
</div>                                                              


<div class="alert alert-block alert-warning">
<b>Изменения:</b> Выводы изменены.
<div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: модели обучены и проверены
    - [x]  Данные корректно разбиты на обучающую и валидационную выборки
    - [x]  Модели обучены, предсказания сделаны
    - [x]  Предсказания и правильные ответы на валидационной выборке сохранены
    - [x]  На экране напечатаны результаты
    - [x]  Сделаны выводы
- [x]  Выполнен шаг 3: проведена подготовка к расчёту прибыли
    - [x]  Для всех ключевых значений созданы константы Python
    - [x]  Посчитано минимальное среднее количество продукта в месторождениях региона, достаточное для разработки
    - [x]  По предыдущему пункту сделаны выводы
    - [x]  Написана функция расчёта прибыли
- [x]  Выполнен шаг 4: посчитаны риски и прибыль
    - [x]  Проведена процедура *Bootstrap*
    - [x]  Все параметры бутстрепа соответствуют условию
    - [x]  Найдены все нужные величины
    - [x]  Предложен регион для разработки месторождения
    - [x]  Выбор региона обоснован